File: src/ipc/ai.ts
/**
 * Typed IPC wrapper for the AI provider registry
 * (Phase 5a), the streaming chat proxy
 * (Phase 5b-1 / 5b-2), and the tool-call
 * surface (Phase 5b-4).
 *
 * Mirrors `src-tauri/src/ai.rs` and
 * `src-tauri/src/chat.rs`. Components import
 * from `@/ipc`, never from `@tauri-apps/api/core`
 * directly (Rule 4).
 *
 * The 5a surface is read-only: a static list of
 * providers and a cheap "which of the 3 have keys"
 * check. The 5b-1 / 5b-2 additions added the
 * streaming chat invoke + cancellation + the
 * three `ai://*` event subscriptions. The 5b-4
 * additions extended the `ChatChunkPayload`
 * union with a `toolCall` variant for
 * function-calling events.
 *
 * The Rust side serialises payloads as camelCase
 * discriminated unions with a `kind` tag.
 */

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';

/**
 * The static list of supported AI providers, as
 * returned by `ai_list_providers`. The Rust side
 * hardcodes the provider metadata (id, base URLs,
 * default model, available models, key URL); the
 * frontend renders one Settings card per entry.
 *
 * The two `*_compatible_base_url` fields are stored
 * but unused in 5a; 5b uses them to route the
 * streaming request. The frontend can ignore them
 * for the Settings UI.
 */
export interface ProviderInfo {
  /** Stable id (e.g. "openai"). Used as the keychain
   *  user-name and the `provider` parameter to AI
   *  IPC commands. Never rename without a key
   *  migration. */
  id: string;
  /** Human-readable display name (e.g. "OpenAI"). */
  displayName: string;
  /** OpenAI-compatible `/v1/chat/completions` base
   *  URL. `null` if the provider doesn't speak it. */
  openaiCompatibleBaseUrl: string | null;
  /** Anthropic-compatible `/v1/messages` base URL.
   *  `null` if the provider doesn't speak it. */
  anthropicCompatibleBaseUrl: string | null;
  /** Default model id for chat completions from
   *  this provider. */
  defaultModel: string;
  /** Hardcoded model list shown in the UI picker. */
  availableModels: string[];
  /** Short blurb shown on the Settings card. */
  description: string;
  /** Link to the provider's API-key page. */
  keyUrl: string;
}

/**
 * Returns the static list of supported AI providers.
 * The Settings screen calls this on mount. The list
 * never changes between Tauri restarts, so the result
 * is safe to cache in a Zustand store and re-read
 * only on screen re-mount.
 */
export async function aiListProviders(): Promise<ProviderInfo[]> {
  return invoke<ProviderInfo[]>('ai_list_providers');
}

/**
 * Returns the ids of providers that have a key in
 * the keychain. Used by the Settings screen to render
 * the "Configured" / "Not configured" badges without
 * three separate `secretsHasApiKey` round-trips.
 *
 * On keychain error for a specific provider, that
 * provider is silently omitted from the result. The
 * Settings screen will surface the error in detail
 * when the user clicks that card (a separate
 * `secretsHasApiKey` call).
 */
export async function aiGetConfiguredProviders(): Promise<string[]> {
  return invoke<string[]>('ai_get_configured_providers');
}

// --- Phase 5b-3: streaming chat IPC --------------------------------------
//
// The streaming chat is an event-driven pipeline,
// not a request/response:
//   1. JS calls `aiChatStream(...)` which returns
//      a `requestId` synchronously.
//   2. The Rust side spawns a task that emits
//      `ai://chunk` events tagged with that
//      `requestId` (1+ per assistant token).
//   3. When the stream is over (or cancelled or
//      errored), the Rust side emits `ai://done`
//      (always) and optionally `ai://error`
//      (on early failure).
//   4. JS subscribes to the three event names
//      ONCE at app startup (in `aiStore.ts`) and
//      demuxes by `requestId`.
//
// The JS side NEVER has to know about SSE
// framing, transport, or cancellation tokens —
// those are all in `src-tauri/src/chat.rs`.

/**
 * A single message in the chat thread. The shape
 * mirrors `ChatMessage` in `src-tauri/src/chat.rs`.
 *
 * `name` is optional (used by some providers for
 * multi-user chats; typically absent). The Rust
 * side accepts it but OpenAI ignores it; Anthropic
 * errors if you set a `name` and it doesn't match
 * its conventions.
 *
 * 5b-6 additions: `toolCalls` and `toolCallId` for
 * the tool-execution loop. The model emits tool
 * calls (assistant messages carry the `toolCalls`
 * array); the JS-side registry executes each tool
 * and sends the result back as a follow-up message
 * with `role: 'tool'` and `toolCallId: <the original
 * call's id>`. The Rust side converts this internal
 * shape into the provider's wire format (OpenAI:
 * direct passthrough; Anthropic: assistant tool
 * calls → `tool_use` content blocks, tool result
 * → `user` role with `tool_result` content blocks).
 */
export interface ChatMessageArgs {
  role: 'system' | 'user' | 'assistant' | 'tool';
  content: string;
  name?: string;
  /**
   * 5b-6: present on assistant messages that
   * emitted one or more tool calls. The order
   * matches the order the model decided to
   * invoke the tools in. Each entry's `id`
   * is the provider-assigned id; the
   * `toolCallId` on the corresponding result
   * message echoes this back.
   */
  toolCalls?: AssistantToolCallArgs[];
  /**
   * 5b-6: present on tool result messages
   * (`role: 'tool'`). The id of the call this
   * is the result of — the JS side reads it
   * from the original `ToolCallPayload.id`.
   */
  toolCallId?: string;
}

/**
 * 5b-6: a tool call attached to an assistant
 * message. Mirrors `AssistantToolCall` in
 * `src-tauri/src/chat.rs`. The JS side sends
 * these back in follow-up requests so the
 * provider knows "this assistant message had
 * these tool calls" (the model uses the id to
 * map a result message back to the call).
 */
export interface AssistantToolCallArgs {
  /** Provider-assigned id. OpenAI: `call_…`,
   *  Anthropic: `toolu_…`. */
  id: string;
  /** Function name, e.g. `'get_file_contents'`. */
  name: string;
  /** Concatenated JSON argument string. May
   *  be `''` for no-arg tools; may not be
   *  valid JSON if the model hallucinated. */
  arguments: string;
}

/**
 * A user-defined custom tool (5c). The JS
 * `customToolsStore` is the source of truth at
 * runtime; the `customTools` field on
 * `ChatStreamArgs` carries a denormalised
 * snapshot of the current contents of
 * `lipi-tools.json` to the Rust side so the
 * model can see the custom tool definitions.
 *
 * Mirrors `CustomToolSpec` in
 * `src-tauri/src/custom_tool.rs`. The two
 * sides are kept in sync by hand for the MVP.
 *
 * 5c scope: only `'shell' | 'http'` tool
 * kinds. The `args` list is a tiny type
 * system — only `'string'` for 5c. Future
 * `'number' | 'boolean' | 'enum'` arg types
 * are 5d+ additions.
 */
export interface CustomToolArg {
  /** Argument name (e.g. `'path'`). Must
   *  match a `{name}` placeholder in the
   *  tool's command/url template.
   *  Case-sensitive. The `customToolsStore`
   *  validates identifier shape on save. */
  name: string;
  /** Argument type. Only `'string'` in 5c. */
  type: 'string';
  /** Human-readable description shown to
   *  the model in the tool's JSON Schema. */
  description: string;
}

/**
 * The wire shape of a custom tool as it
 * crosses the JS↔Rust boundary. The Rust
 * side doesn't see the tool's
 * `command`/`url`/etc. (only the JS
 * executor needs those) — it only needs
 * enough to build the provider-specific
 * tool declaration. The full tool
 * configuration (including the
 * command/url/headers) lives in
 * `lipi-tools.json` and is loaded by the
 * `customToolsStore` at app start.
 */
export interface CustomToolSpec {
  /** Tool name (e.g. `'run_npm_test'`).
   *  Must be unique across the registry.
   *  The `customToolsStore` enforces this
   *  on save. */
  name: string;
  /** Human-readable description shown to
   *  the model in the tool's JSON Schema. */
  description: string;
  /** Ordered list of arguments. The order
   *  is preserved in the generated JSON
   *  Schema (matters for the Anthropic
   *  tool spec). */
  args: CustomToolArg[];
}

/**
 * The request body for `ai_chat_stream`. Pass the
 * full thread (NOT just the new user message) —
 * the Rust side is stateless about the thread;
 * the `aiStore` is the source of truth.
 *
 * 5b-7 additions:
 *   - `enabledToolNames` — per-tool enable/disable
 *     whitelist. The JS `toolSettingsStore` is the
 *     source of truth (the user toggles tools on
 *     the Settings screen); the AI store snapshots
 *     the current state on every `send()` and passes
 *     it through. The Rust side uses it to filter
 *     the `tools: [...]` array sent to the provider,
 *     so a disabled tool is invisible to the model.
 *     An absent / empty array means "all enabled"
 *     (the legacy default — see `chat.rs`'s
 *     `is_tool_enabled` for the exact semantics).
 *
 * 5c additions:
 *   - `customTools` — denormalised snapshot of the
 *     current `customToolsStore` state. The Rust
 *     side merges these on top of the built-in
 *     `TOOL_CATALOGUE` and shapes them per-provider
 *     (OpenAI: `{type:"function", function:{…}}`,
 *     Anthropic: `{name, description, input_schema}`).
 *     The actual EXECUTION of a custom tool happens
 *     on the JS side (in `toolRegistry`); the Rust
 *     side only sees the shape definition.
 */
export interface ChatStreamArgs {
  /** One of `'openai' | 'anthropic' | 'openrouter'`. */
  provider: string;
  /**
   * Model id (e.g. `'gpt-4o-mini'`). Optional — if
   * omitted, the Rust side uses the provider's
   * `defaultModel` from `aiListProviders()`.
   */
  model?: string;
  /** The full chat thread, oldest first. */
  messages: ChatMessageArgs[];
  /**
   * 5b-7: names of tools the model is allowed
   * to call. The JS `toolSettingsStore` collects
   * this from the user; the AI store snapshots
   * it on every `send()` and passes it through.
   * An empty array is the "all enabled" default
   * (backwards-compat for code paths that don't
   * know about 5b-7 yet — the Rust side treats
   * this as "no filter"). The wire field is
   * `enabledToolNames` (camelCase); the Rust
   * struct is `enabled_tool_names`.
   */
  enabledToolNames?: string[];
  /**
   * 5c: per-request custom tool list. The JS
   * `customToolsStore` is the source of truth
   * for which custom tools exist; this is a
   * denormalised snapshot of the current
   * `lipi-tools.json` contents. Empty /
   * undefined = no custom tools (the default
   * for code paths that don't know about 5c
   * yet). The `enabledToolNames` filter still
   * applies — a custom tool whose name is not
   * in the whitelist is invisible to the
   * model. The wire field is `customTools`
   * (camelCase); the Rust struct is
   * `custom_tools`.
   */
  customTools?: CustomToolSpec[];
}

/**
 * A complete tool call (5b-4). The model has
 * decided to invoke the named function with
 * the given JSON argument. Mirrors
 * `ChatDelta::ToolCall` on the Rust side.
 *
 * The Rust adapter assembles the `input` JSON
 * byte-by-byte from the provider's streaming
 * protocol — by the time we receive this
 * payload, `input` is the fully concatenated
 * JSON string. We do NOT parse it on the
 * Rust side; the JS side may want to display
 * it raw (for transparency) or parse it
 * (for execution).
 *
 * The `id` is provider-assigned (OpenAI:
 * `call_…`, Anthropic: `toolu_…`). The JS
 * side uses this as the React `key` for
 * the per-message tool trace and (in a
 * future phase) as the id to map a
 * follow-up "tool result" message back to
 * the original call.
 */
export interface ToolCallPayload {
  kind: 'toolCall';
  /** Provider-assigned id. */
  id: string;
  /** Function name, e.g. `'get_weather'`. */
  name: string;
  /**
   * Concatenated JSON argument string. May
   * be `''` for no-arg tools, or invalid JSON
   * if the model hallucinated. The JS side
   * should `JSON.parse` with a try/catch
   * when it wants to execute the tool.
   */
  input: string;
}

/**
 * The `ai://chunk` event payload. The `payload`
 * field is a discriminated union with a `kind`
 * tag — the same shape `ChatDelta` has on the
 * Rust side.
 *
 * 5b-4 added `ToolCall` to the union. The
 * store's demux appends `toolCall` to the
 * current streaming assistant message's
 * `toolCalls` array.
 */
export type ChatChunkPayload =
  | { kind: 'delta'; text: string }
  | ToolCallPayload
  | {
      kind: 'done';
      cancelled: boolean;
      /**
       * Anthropic only. One of
       * `'end_turn' | 'max_tokens' | 'stop_sequence' | 'tool_use'`.
       * `undefined` for OpenAI / OpenRouter and for
       * user-cancelled completions. `'tool_use'`
       * means the model finished its turn by
       * emitting one or more tool calls; the
       * corresponding `toolCall` chunks will
       * have arrived BEFORE this `done` chunk
       * (5b-4).
       */
      stopReason?: string;
    }
  | { kind: 'error'; errorKind: string; message: string };

/**
 * Envelope for every `ai://chunk` event. The
 * `requestId` is the same one returned by
 * `aiChatStream()`; the JS store demuxes by it.
 */
export interface ChunkEnvelope {
  requestId: string;
  payload: ChatChunkPayload;
}

/**
 * Envelope for the `ai://done` event. This is the
 * authoritative "stream is over" signal for the
 * JS store (separate from the per-chunk `done`
 * payload above, which is the inline-display
 * signal). Both fire on completion; the JS store
 * treats them as the same event.
 */
export interface DoneEnvelope {
  requestId: string;
  cancelled: boolean;
  /** See `ChatChunkPayload`'s `done.stopReason`. */
  stopReason?: string;
}

/**
 * Envelope for the `ai://error` event. Fires on
 * early failure (before the first chunk) — e.g.
 * the keychain read failed, the provider is
 * unknown, the request never made it out. Errors
 * that happen mid-stream surface as
 * `ai://chunk` events with `kind: 'error'`
 * instead.
 */
export interface ErrorEnvelope {
  requestId: string;
  kind: string;
  message: string;
}

/**
 * Open a streaming chat completion. Returns the
 * `requestId` synchronously so the caller can
 * subscribe to `ai://chunk` / `ai://done` /
 * `ai://error` BEFORE the first chunk arrives
 * (the provider's first event can come back in
 * <50ms; the JS subscription must be in place
 * first, or the chunk is lost).
 *
 * `aiChatStream` does NOT throw on the happy
 * path — it returns a `requestId` even if the
 * provider is mid-error. The Rust side surfaces
 * provider errors as `ai://error` events (or
 * `ai://chunk` events with `kind: 'error'`).
 *
 * Throws ONLY for setup failures the JS side
 * can pre-validate (e.g. Tauri command not
 * registered, IPC channel closed).
 */
export async function aiChatStream(args: ChatStreamArgs): Promise<string> {
  return invoke<string>('ai_chat_stream', { args });
}

/**
 * Cancel an in-flight chat stream. The Rust side
 * flips the cancel flag; the reader task emits
 * `ai://done` with `cancelled: true` on the next
 * SSE event check.
 *
 * Returns `true` if the request was found and
 * cancelled; `false` if it was already gone
 * (finished naturally before the user clicked
 * Stop, or the requestId is unknown). The JS
 * store treats `false` as a no-op.
 */
export async function aiCancelStream(requestId: string): Promise<boolean> {
  return invoke<boolean>('ai_cancel_stream', { requestId });
}

/**
 * Subscribe to `ai://chunk` events. Returns an
 * unlisten function. The store calls this ONCE at
 * app startup and demuxes by `requestId` in the
 * callback.
 */
export async function onAiChunk(
  cb: (envelope: ChunkEnvelope) => void,
): Promise<UnlistenFn> {
  return listen<ChunkEnvelope>('ai://chunk', (e) => cb(e.payload));
}

/**
 * Subscribe to `ai://done` events. The store
 * listens for these to clear the
 * "streaming…" status and persist the final
 * message.
 */
export async function onAiDone(
  cb: (envelope: DoneEnvelope) => void,
): Promise<UnlistenFn> {
  return listen<DoneEnvelope>('ai://done', (e) => cb(e.payload));
}

/**
 * Subscribe to `ai://error` events. Fires on
 * early failure (before the first chunk). The
 * store surfaces these as inline errors in the
 * chat thread.
 */
export async function onAiError(
  cb: (envelope: ErrorEnvelope) => void,
): Promise<UnlistenFn> {
  return listen<ErrorEnvelope>('ai://error', (e) => cb(e.payload));
}

File: src-tauri/src/ai.rs
//! AI provider registry + streaming proxy to LLM providers.
//!
//! Phase 5a: static provider table, `list_providers`,
//! `get_configured_providers`. No network calls.
//!
//! Phase 5b-1 (this phase): OpenAI-compatible streaming
//! chat. One provider adapter, one SSE parser, one Tauri
//! command. The OpenRouter adapter is a base-URL swap and
//! lives in 5b-2 alongside the Anthropic adapter. The
//! 5b-1 surface is:
//!
//!   - `ChatMessage`, `ChatDelta`, `ChatError` types
//!     (serialised to camelCase JSON for the JS side).
//!   - `SseStream` — a tiny SSE parser over an
//!     `AsyncRead`. Handles `data: {json}\n\n` framing,
//!     `[DONE]` sentinels, partial UTF-8 across chunks,
//!     and comment lines. Pure byte-IO; doesn't depend
//!     on reqwest (so the tests can feed it a `Cursor`).
//!   - `stream_chat_openai(api_key, base_url, model,
//!     messages, on_chunk, cancel) -> Result<(), ChatError>`.
//!     Opens an HTTPS POST, parses the SSE body, and
//!     invokes `on_chunk` for each `Delta` / `Done` /
//!     `Error`. The `cancel` token is checked between
//!     chunks; when flipped, the function returns
//!     `Ok(())` after emitting a synthetic
//!     `Done { cancelled: true }` chunk.
//!   - `#[tauri::command] fn ai_chat_stream(...)` —
//!     reads the API key from the keychain, generates
//!     a `requestId`, spawns a tokio task that calls
//!     `stream_chat_openai` and emits `ai://chunk` /
//!     `ai://done` / `ai://error` events to the main
//!     window. Returns the `requestId` synchronously
//!     so the JS side can subscribe before the first
//!     chunk arrives (same pattern as 4a's terminal).
//!
//! Phase 5b-2 will add:
//!   - OpenRouter adapter (base-URL swap of OpenAI).
//!   - Anthropic adapter (different SSE framing, named
//!     events `message_start` / `content_block_delta` /
//!     `message_stop`; different request body and
//!     auth headers).
//!   - `#[tauri::command] fn ai_cancel_stream(...)` —
//!     flips the cancellation token in a
//!     `HashMap<requestId, Arc<AtomicBool>>` so the
//!     reader task bails out.
//!
//! ## Why a static table (and not a live `/models` fetch)
//!
//! Fetching the live model list per-provider is doable
//! (OpenAI's `/v1/models`, Anthropic's `/v1/models`, etc.)
//! but adds a network call to every Settings screen open
//! and surfaces provider-specific quirks (OpenRouter has
//! hundreds of models, Anthropic paginates). For 5a the
//! curated list of 6 models (2 per provider) is enough.
//! 5b-or-later can add a "Refresh models" button that
//! hits the live endpoint and merges the results.
//!
//! ## Per-provider base URLs and auth headers
//!
//! All three providers use Bearer-token auth. The base URL
//! is per-provider. OpenAI and OpenRouter use the same
//! `/v1/chat/completions` endpoint and the same
//! streaming-event format (so OpenRouter is a base-URL
//! swap, not a separate adapter). Anthropic uses
//! `/v1/messages` and a different event format
//! (`message_start` / `content_block_delta` / etc.).

use serde::Serialize;

use crate::secrets;

/// A static description of a supported AI provider.
/// Exposed to the frontend via `ai_list_providers`.
/// The `id` is the user name in the OS keychain
/// (`service = "app.lipi.ide"`, `user = <id>`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderInfo {
    /// Stable id, used in the keychain and as the
    /// "provider" parameter to all AI IPC commands.
    /// Never rename without a migration.
    pub id: &'static str,
    /// Human-readable display name for the UI.
    pub display_name: &'static str,
    /// Base URL for the OpenAI-compatible
    /// `/v1/chat/completions` endpoint. `None`
    /// for providers (Anthropic) that need a
    /// different endpoint and a dedicated adapter.
    /// 5a: stored but unused; 5b uses it.
    pub openai_compatible_base_url: Option<&'static str>,
    /// Base URL for the Anthropic-compatible
    /// `/v1/messages` endpoint. Symmetric to
    /// `openai_compatible_base_url`. `None` for
    /// providers that don't speak it.
    /// 5a: stored but unused; 5b uses it.
    pub anthropic_compatible_base_url: Option<&'static str>,
    /// The default model for chat completions
    /// from this provider. The user can override
    /// per-request via the `model` field in 5b.
    pub default_model: &'static str,
    /// The hardcoded model list shown in the UI
    /// picker. We expose 2 per provider (one fast
    /// and cheap, one capable) for 5a.
    pub available_models: &'static [&'static str],
    /// Short blurb shown on the Settings card.
    pub description: &'static str,
    /// Link to the provider's API key page. Shown
    /// on the Settings card as "Get a key →".
    pub key_url: &'static str,
}

/// Return the static list of supported providers. The
/// 3 providers are ordered: OpenAI first (most common),
/// Anthropic second (most capable), OpenRouter third
/// (unified access). Re-ordering in the UI is fine;
/// this order is the canonical "first" order.
pub fn list_providers() -> Vec<ProviderInfo> {
    vec![
        ProviderInfo {
            id: "openai",
            display_name: "OpenAI",
            openai_compatible_base_url: Some("https://api.openai.com/v1"),
            anthropic_compatible_base_url: None,
            default_model: "gpt-4o-mini",
            available_models: &["gpt-4o-mini", "gpt-4o"],
            description: "GPT-4o and GPT-4o mini. Fast, capable, widely used.",
            key_url: "https://platform.openai.com/api-keys",
        },
        ProviderInfo {
            id: "anthropic",
            display_name: "Anthropic",
            openai_compatible_base_url: None,
            anthropic_compatible_base_url: Some("https://api.anthropic.com/v1"),
            default_model: "claude-3-5-haiku-latest",
            available_models: &["claude-3-5-haiku-latest", "claude-3-5-sonnet-latest"],
            description: "Claude 3.5 Haiku and Sonnet. Strong at code and reasoning.",
            key_url: "https://console.anthropic.com/settings/keys",
        },
        ProviderInfo {
            id: "openrouter",
            display_name: "OpenRouter",
            openai_compatible_base_url: Some("https://openrouter.ai/api/v1"),
            anthropic_compatible_base_url: None,
            default_model: "anthropic/claude-3.5-sonnet",
            available_models: &[
                "anthropic/claude-3.5-sonnet",
                "openai/gpt-4o-mini",
            ],
            description: "Unified access to many models. One key, many providers.",
            key_url: "https://openrouter.ai/keys",
        },
    ]
}

/// Look up a provider by id. Returns `None` if the id
/// is not in the static list. Used by 5b to validate
/// the `provider` field of an `ai_chat_stream` call.
pub fn provider_by_id(id: &str) -> Option<ProviderInfo> {
    list_providers().into_iter().find(|p| p.id == id)
}

/// Return the ids of providers that have a key in the
/// keychain. Used by the AI panel to render the
/// "configured" / "not configured" badges without
/// making the frontend call `has_api_key` 3 times.
///
/// If a keychain error occurs for one provider, that
/// provider is silently omitted from the result
/// (the frontend's separate `has_api_key` call will
/// surface the error in detail when the user clicks
/// that card). We never propagate keychain errors
/// out of this "list the configured ones" helper —
/// it's a best-effort UI hint, not a security check.
pub fn get_configured_providers(
    snapshot_path: Option<&std::path::Path>,
) -> Vec<&'static str> {
    list_providers()
        .iter()
        .filter_map(|p| {
            match secrets::has_api_key(p.id, snapshot_path) {
                Ok(true) => Some(p.id),
                _ => None,
            }
        })
        .collect()
}

// --- Phase 5b-1: streaming chat proxy -------------------------------------
//
// The actual SSE parser and OpenAI provider adapter
// live in `src/chat.rs` (Rule 3 — single file,
// single concern). The `mod chat;` declaration is
// in `lib.rs` so the file lives at the crate root.

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn list_providers_returns_three() {
        let providers = list_providers();
        assert_eq!(providers.len(), 3);
        let ids: Vec<&str> = providers.iter().map(|p| p.id).collect();
        assert_eq!(ids, vec!["openai", "anthropic", "openrouter"]);
    }

    #[test]
    fn list_providers_have_required_fields() {
        for p in list_providers() {
            assert!(!p.id.is_empty());
            assert!(!p.display_name.is_empty());
            assert!(!p.default_model.is_empty());
            assert!(!p.available_models.is_empty());
            assert!(p.available_models.contains(&p.default_model));
            assert!(!p.description.is_empty());
            assert!(!p.key_url.is_empty());
            assert!(p.key_url.starts_with("https://"));
        }
    }

    #[test]
    fn provider_by_id_finds_known_providers() {
        assert_eq!(provider_by_id("openai").unwrap().id, "openai");
        assert_eq!(provider_by_id("anthropic").unwrap().id, "anthropic");
        assert_eq!(provider_by_id("openrouter").unwrap().id, "openrouter");
    }

    #[test]
    fn provider_by_id_returns_none_for_unknown() {
        assert!(provider_by_id("grok").is_none());
        assert!(provider_by_id("").is_none());
        assert!(provider_by_id("OpenAI").is_none()); // case-sensitive
    }

    #[test]
    fn get_configured_providers_empty_when_no_keys() {
        // No keys in the keychain (the test runner
        // uses the mock store; we don't preset any
        // passwords). Should be empty.
        // Note: this test depends on secrets using
        // the mock builder. We can't set the mock
        // builder from here (it would race with the
        // secrets.rs tests' Once). For the test
        // runner, the secrets tests install the mock
        // and run first; if they don't, this test
        // is best-effort and may report "all three
        // configured" — which is also a valid
        // pass-through for the *shape* of the result.
        // We assert just that the result is a subset
        // of known providers.
        let configured = get_configured_providers(None);
        for id in &configured {
            assert!(provider_by_id(id).is_some());
        }
    }
}
